docs/ABI/TypeMetadata.rst

```diff
@@ -477,3 +509,236 @@ contains:
+  initialization that is required in order to support basic operations
+  on the type.  For example, a metadata in this state will have undergone
+  any necessary "internal layout" that might be required in order to
+  create values of the type but not to allocate storage to hold them.
+  For example, a class metadata will have an instance layout, which is
+  not required in order to compute the external layout, but is required
+  in order to allocate instances or create a subclass.
+
+- A **complete** metadata additionally makes certain guarantees of
+  transitive completeness of the metadata referenced from the metadata.
+  For example, a complete metadata for ``Array<T>`` guarantees that
+  the metadata for ``T`` stored in the generic arguments vector is also
+  complete.
+
+Metadata never backtrack in their state.  In particular, once metadata
+is complete, it remains complete forever.
+
+Transitive Completeness Guarantees
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A complete class metadata makes the following guarantees:
+
+- Its superclass metadata (if it has a superclass) is complete.
+- Its generic arguments (if it has any) are complete.
+- By implication, the generic arguments of its superclasses are complete.
+
+A complete struct, enum, or optional metadata makes the following guarantees:
+
+- Its generic arguments (if it has any) are complete.
+
+A complete tuple metadata makes the following guarantees:
+
+- Its element types are complete.
+
+Other kinds of type metadata do not make any completeness guarantees.
+The metadata kinds with transitive guarantees are the metadata kinds that
+potentially require two-phase initialization anyway.  Other kinds of
+metadata could otherwise declare themselves complete immediately on
+allocation, so the transitive completeness guarantee would add significant
+complexity to both the runtime interface and its implementation, as well
+as adding probably-unrecoverable memory overhead to the allocation process.
+
+It is also true that it is far more important to be able to efficiently
+recover complete metadata from the stored arguments of a generic type
+than it is to be able to recover such metadata from a function metadata.
+
+Completeness Requirements
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Type metadata are required to be transitively complete when they are
+presented to most code.  This allows that code to work with the metadata
+without explicitly checking for its completeness.  Metadata in the other
+states are typically encountered only when initializing or building up
+metadata.
+
+Specifically, a type metadata record is required to be complete when:
+
+- It is passed as a generic argument to a function (other than a metadata
+  access function, witness table access function, or metadata initialization
+  function).
+- It is stored as a metatype value.
+- It is used to build an existential value.
+- It is passed as the ``Self`` argument to a ``static`` or ``class`` method,
```

Comment:
What about when it is passed as a non-self metatype argument?

Reply:
"It is stored as a metatype value".  Maybe "stored" is too leading.